pkg/operator/ceph/cluster/mon/mon.go

```diff
@@ -426,9 +426,28 @@ func (c *Cluster) startDeployments(mons []*monConfig, requireAllInQuorum bool) e
 		if err != nil {
 			return fmt.Errorf("failed to create mon %s. %+v", mons[i].DaemonName, err)
 		}
+		// For the initial deployment (first creation) it's expected to not have all the monitors in quorum
+		// However, in an event of an update, it's crucial to proceed monitors by monitors
+		// At the end of the method we perform one last check where all the monitors must be in quorum
+		requireAllInQuorum := false
+		err = c.waitForMonsToJoin(mons, requireAllInQuorum)
+		if err != nil {
+			return fmt.Errorf("failed to check mon quorum %s. %+v", mons[i].DaemonName, err)
+		}
 	}
 
 	logger.Infof("mons created: %d", len(mons))
+	// Final verification that **all** mons are in quorum
+	// Do not proceed if one monitor is still syncing
+	// Only do this when monitors versions are different so we don't block the orchestration if a mon is down.
+	versions, err := client.GetCephVersions(c.context)
+	if err != nil {
+		logger.Warningf("failed to get ceph daemons versions. %+v, this likely means there is no cluster yet.", err)
+	} else {
+		if len(versions.Mon) != 1 {
```

Comment:
Remind me why it isn't sufficient to query if the mons are ok-to-stop when they are being upgraded? For example, if there are 3 mons, is the following sequence valid?
1. Wait for mon A to be ok-to-stop
1. Upgrade mon A
1. Wait for the mon A pod to start
1. Return to step 1 for mon B, then mon C

When we wait for mon B to be ok-to-stop in step 1, won't that already fulfill the requirement that mon A is started and we have sufficient quorum before mon B is stopped? Just trying to confirm why we need to add these checks for being in quorum.

Reply:
What you're saying is correct but enforcing monitors to be in quorum here allows a smoother `ok-to-stop`, basically, if we had 3 healthy mons and we upgrade them, they will always be ok-to-stop thanks to that check. If we don't force this here, the orchestration update will loop and fail. I've found it better to avoid sending errors messages back to the operator's log if possible, it's less confusing for users.

In this case, the ok-to-stop check might fail if we already have 2 out of 3 mons up for example.

Makes sense?

Reply:
I was thinking the OkToStop would retry until the daemon is ready to be stopped. Now that I look again, it seems the retry only happens for OkToContinue, but not OkToStop. Should we also retry for OkToStop? Seems like that would help reduce the errors when it's just taking a couple minutes during upgrade before a daemon is ready, and perhaps that removes this need for waiting for quorum?

Reply:
What is the reasoning/resolution to this?

Reply:
The resolution is to always enforce monitor being in quorum during an upgrade.